### PR TITLE
Replace WrapOptions with Into<Options>

### PR DESCRIPTION
**Pull request by Koxiaet**
It's fairly confusing to have a struct and a nearly-identical trait. By centering everything around the concrete `Options` type, the relationship between `usize` and `Options` can be made more clear.

### CODE REVIEW

**Comment by mgeisler** [comment]
Hey @Koxiaet, this is cool! Thank you — i agree that this makes things much clearer :-)

Can I get you to take out unrelated changes from your commits and also to squash the two commits into a single self-contained commit? I'll review the change now and try to point out what I mean.

File: src/core.rs [hunk 1]
```diff
@@ -124,7 +125,7 @@ impl<'a> Word<'a> {
                 if width > 0 && width + ch_width > line_width {
                     let word = Word {
                         word: &self.word[offset..idx],
-                        width: width,
```
Comment (mgeisler) [inline src/core.rs:127]: This change and the other similar changes are unrelated.

File: src/lib.rs [hunk 2]
```diff
@@ -106,27 +106,8 @@ pub use crate::splitting::{HyphenSplitter, NoHyphenation, WordSplitter};
...
-    /// Allow long words to be broken if they cannot fit on a line.
-    /// When set to `false`, some lines may be longer than `width`.
-    fn break_words(&self) -> bool;
-    /// Find indices where `word` can be split.
-    fn split_points(&self, word: &str) -> Vec<usize>;
-}
-
 /// Holds settings for wrapping and filling text.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Copy)]
```
Comment (mgeisler) [inline src/lib.rs:110]: Is this related somehow? As far as i recall, the `Standard` type from the `hyphenation` crate is not `Copy`, so I guess the derived copy implementation won't work then — is there then a reason to have it?

File: src/lib.rs [hunk 3]
```diff
@@ -144,79 +125,27 @@ pub struct Options<'a, S = Box<dyn WordSplitter>> {
...
-        self.subsequent_indent
-    }
-    #[inline]
-    fn break_words(&self) -> bool {
-        self.break_words
-    }
-    #[inline]
-    fn split_points(&self, word: &str) -> Vec<usize> {
-        self.splitter.split_points(word)
+impl<'a, S> From<&'a Options<'a, S>> for Options<'a, &'a S> {
```
Comment (mgeisler) [inline src/lib.rs:128]: This looks a bit weird to me — it's basically saying that you can go from `&Options` to `Options` easily. This is the same as what `*options` would do, right? So would we avoid it somehow?

File: src/lib.rs [hunk 4]
```diff
@@ -520,7 +449,7 @@ impl<'a, S: WordSplitter> Options<'a, S> {
             initial_indent: self.initial_indent,
             subsequent_indent: self.subsequent_indent,
             break_words: self.break_words,
-            splitter: splitter,
```
Comment (mgeisler) [inline src/lib.rs:523]: Unrelated.

File: src/lib.rs [hunk 5]
```diff
@@ -580,7 +509,7 @@ pub fn termwidth() -> usize {
 /// ```
 ///
 /// [`wrap`]: fn.wrap.html
-pub fn fill<T: WrapOptions>(text: &str, options: T) -> String {
+pub fn fill<'a, S: WordSplitter, T: Into<Options<'a, S>>>(text: &str, options: T) -> String {
```
Comment (mgeisler) [inline src/lib.rs:512]: Is the `'a` lifetime related to the elided lifetime for `text`? Could `'a` be turned into `'_` perhaps?

File: src/lib.rs [hunk 6]
```diff
@@ -1149,15 +1078,14 @@ mod tests {
         let green_hello = "\u{1b}[0m\u{1b}[32mHello\u{1b}[0m";
         let blue_world = "\u{1b}[0m\u{1b}[34mWorld!\u{1b}[0m";
         assert_eq!(
-            fill(&(String::from(green_hello) + " " + &blue_world), 6),
-            String::from(green_hello) + "\n" + &blue_world
+            fill(&format!("{} {}", green_hello, blue_world), 6),
```
Comment (mgeisler) [inline src/lib.rs:1081]: While this is a nice change, it seems unrelated. It would be great to have this as a esparate PR.

File: src/lib.rs [hunk 7]
```diff
@@ -1149,15 +1078,14 @@ mod tests {
...
+            format!("{}\n{}", green_hello, blue_world),
         );
     }
 
     #[test]
     fn cloning_works() {
         static OPT: Options<HyphenSplitter> = Options::with_splitter(80, HyphenSplitter);
-        let opt = OPT.clone();
```
Comment (mgeisler) [inline src/lib.rs:1160]: I don't think the test is doing any cloning now?

File: src/splitting.rs [hunk 8]
```diff
@@ -47,6 +47,11 @@ impl<S: WordSplitter + ?Sized> WordSplitter for Box<S> {
     }
 }
 */
+impl<T: WordSplitter> WordSplitter for &T {
```
Comment (mgeisler) [inline src/splitting.rs:50]: This is another funny looking construct :-) Is this perhaps a use case for the [`Borrow`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html#examples) stdlib trait?

What I mean is that the function that have a type bound on `WordSplitter` could perhaps use a bound which allows them to accept both a `WordSplitter` and a borrowed `WordSplitter`.

**Review by mgeisler — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (Koxiaet) [reply src/lib.rs:110]: The reasoning was that it allows `Config` to be `Copy` in many places, which avoids users having to pass references to it when a copy would be fine.

File: src/lib.rs (hunk 3 above)
Comment (Koxiaet) [reply src/lib.rs:128]: It's different because it also takes a reference to `S` which removes the need for it to be `Clone`. I also wanted to support passing `&Config` to the functions as it avoids breaking `split_words` too much, which used to take an `&impl WrapOptions`.

File: src/lib.rs (hunk 5 above)
Comment (Koxiaet) [reply src/lib.rs:512]: Unfortunately not, Rust requires explicit lifetimes here.

File: src/splitting.rs (hunk 8 above)
Comment (Koxiaet) [reply src/splitting.rs:50]: It's a fairly common pattern - see [`AsRef`](https://doc.rust-lang.org/stable/std/convert/trait.AsRef.html#impl-AsRef%3CU%3E), [`Display`](https://doc.rust-lang.org/stable/std/fmt/trait.Display.html#impl-Display-82), [`Hash`](https://doc.rust-lang.org/stable/std/hash/trait.Hash.html#impl-Hash-59), [`Fn`](https://doc.rust-lang.org/stable/std/ops/trait.Fn.html#impl-Fn%3CA%3E). It's also required for the implementation of `From<&Options<'_, S>>` for `Options<'_, &S>` (notice it produces an `&S`, which allows `S` to not be `Clone`).

File: src/lib.rs (hunk 3 above)
Comment (mgeisler) [reply src/lib.rs:128]: Ah, I hadn't spotted the `S` to `&'a S` change :-)

As for breaking `split_words` — feel free to break it all you want, it's a brand new function and I'm happy to hammer out a better signature before I release it to the wild.

File: src/lib.rs (hunk 5 above)
Comment (mgeisler) [reply src/lib.rs:512]: Okay, no worries!

File: src/splitting.rs (hunk 8 above)
Comment (mgeisler) [reply src/splitting.rs:50]: Thanks for those references to the stdlib! This makes me feel like we're in good company :-)

File: src/lib.rs (hunk 3 above)
Comment (Koxiaet) [reply src/lib.rs:128]: `wrap` internally calls `split_words` multiple times. If this was not provided then it would have to (a) clone the `Config` each time and bound `S: Clone` or (b) manually write out the body of this impl, both of which this impl seems preferable to.

That being said, I agree that this impl it is a bit weird. If ergonomics was not an issue all the functions would take `&Options<'_, S>`. However, we want to support `usize`, `Options<'_, S>` _and_ `&Options<'_, S>`. But it's hard to find anything that they all have in common:
* `usize` supports conversion from a reference or value to owned options.
* `Options<'_, S>` supports conversion from a reference to shared options or a value to owned options, or a reference to owned options if we include conversions to `Options<'_, &S>`.
* `&Options<'_, S>` supports conversion from a reference or value to shared options, or to owned options if we include conversions to `Options<'_, &S>`.

The current solution converts a value to owned options, as that's the only thing supported by all three. But you do end up with weird double-references sometimes.

Comment (mgeisler) [reply src/lib.rs:128]: Thanks for that overview, makes this much clearer! It seems I've managed to come up with a bit of a "interesting" API with my weird insistence that `wrap` and `fill` can take both `usize` and some-form-of-options (don't care much here) :smile: 

As for double references... I'm not so concerned with them as long as they're internal to `textwrap` — in particular functions in `textwrap::core` which I consider a place for advanced functionality.

**Comment by mgeisler** [comment]
I'm just trying to understand who the changes here affect the changes in #219 — I hope we can merge both soon and get a nice cleaned up API overall.

File: src/lib.rs (hunk 2 above)
Comment (mgeisler) [reply src/lib.rs:110]: We'll still need to treat it as non-`Copy` internally, since we cannot rely on `Copy` being derived. This makes me think that it should at least go in a separate PR. I'm still a bit wary of offering one API for users who use a `Copy` type as the `WordSplitter` and a different API for other users — it might be surprising to people is their code stops compiling with the `hyphenation` feature. On the other hand, they might never use that feature and then it's indeed nicer for people if they can just copy around their `Options` as they like.

So yeah, please spin this off into a separate PR which we can then merge in parallel with this.

**Review by mgeisler — APPROVED** [review]

**Comment by mgeisler** [comment]
Thanks for the updates and explanations!
